include/point_one/polaris_asio_embedded_client.h

```diff
@@ -0,0 +1,657 @@
+        boost::property_tree::read_json(reply_body, pt);
+        challenge = pt.get<std::string>("challenge");
+        VLOG(3) << "Challenge is : " << challenge;
+      } catch (std::exception &e) {
+        LOG(ERROR) << "Exception in parsing challenge request response: " << e.what();
+        return "";
+      }
+
+      return challenge;
+    } catch (std::exception &e) {
+      LOG(ERROR) << "Exception: " << e.what();
+    }
+    return "";
+  }
+
+  std::string BuildChallengeSignature(std::string challenge, uint64_t time) {
+    // concatenate the string and calculate the message digest
+
+    // challenge| timestamp | company_id | serial_number
+    std::stringstream ss;
+    ss << challenge << std::to_string(time) << company_id_ << unique_id_;
+
+    unsigned int HMAC_SHA512_HASH_LEN = 64;
+    uint8_t buf[HMAC_SHA512_HASH_LEN];
+    std::string message = ss.str();
+    HMAC_CTX hmac_ctx;
+    HMAC_CTX_init(&hmac_ctx);
+    HMAC_Init_ex(&hmac_ctx, (void *)signing_secret_.data(),
+                 signing_secret_.length(), EVP_sha512(), NULL);
+    HMAC_Update(&hmac_ctx, (uint8_t *)message.data(), message.length());
+    HMAC_Final(&hmac_ctx, buf, &HMAC_SHA512_HASH_LEN);
+    unsigned char encodedData[128] = {};  // this is beyond safe, the max length
+                                          // of this is 1.25 * the HASH LEN
+
+    EVP_EncodeBlock(encodedData, buf, HMAC_SHA512_HASH_LEN);
+    std::string output((char *)encodedData);
+    VLOG(2) << "Challenge Response Signature :" << std::endl << output;
+    return output;
+  }
+
+  bool GetAuthToken(std::string challenge, uint64_t time, std::string signature) {
+    try {
+      // Get a list of endpoints corresponding to the server name.
+      boost::asio::ip::tcp::resolver resolver(io_service_);
+      boost::asio::ip::tcp::resolver::query query(connection_settings_.api_host,
+                                                  "http");
+      boost::asio::ip::tcp::resolver::iterator endpoint_iterator =
+              resolver.resolve(query);
+
+      // Try each endpoint until we successfully establish a connection.
+      boost::asio::ip::tcp::socket socket(io_service_);
+      boost::asio::connect(socket, endpoint_iterator);
+
+      // Form the request. We specify the "Connection: close" header so that the
+      // server will close the socket after transmitting the response.
+      boost::asio::streambuf request;
+      std::ostream request_stream(&request);
+
+
+      std::stringstream post_body;
+      post_body << "{\"challenge\": \"" << challenge << "\", "
+                << "\"timestamp\":  \"" << std::to_string(time) << "\", "
+                << "\"company_id\": \"" << company_id_ << "\", "
+                << "\"unique_id\": \"" << unique_id_ << "\", "
+                << "\"signature\": \"" << signature << "\""
+                << "}\r\n";
+      std::string post_body_str = post_body.str();
+      LOG(INFO) << " JSON " << std::endl << post_body_str;
```

Comment:
This should be a low level VLOG, and if you keep it at all it really needs more context.